### ISSUE TITLE
External MIDI players affected by Linux 6.5 kernel regression

### ISSUE THREAD

**Issue by AaronBPaden** [dosbox-staging/dosbox-staging#2816]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

v0.81.0-alpha-1569-g6ce93db22

### What Operating System are you using?

Linux x86_64

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Something funky is going on with dosbox and alsa. I typically build dosbox-staging with fluidsynth turned off, opting instead for a system service because I will occasionally play windows games with midi or listen to midi music as well and I've got the spare RAM.

But I've just noticed that the current dosbox-staging is somehow breaking the service. On a fresh boot, I am able to play midi files. But if I launch Duke Nukem 3d with dosbox, configured for general midi, I get silence. After closing dosbox, I will not be able to use the service again until a reboot (restarting the service did not appear to work).

### Steps to reproduce the behaviour.

Explain how to reproduce
1.  Start a fluidsynth system service using systemd unit files
2.  Launch Duke Nukem 3D, configured with general midi music using the following dosbox.conf

### Your configuration

```ini
https://gist.github.com/AaronBPaden/0e2ef3e4441f33af6b8030b34ad050fe
```

### Provide a Log

```text
https://gist.github.com/AaronBPaden/349b7ddd0e96b7c8a55b9d650148b443
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by AaronBPaden** [comment]
Just tested with the provided build of dosbox-staging 0.80.1 and am seeing the same issue.

**Comment by johnnovak** [comment]
AFAIK, ALSA is heavily deprecated on our side and thus not well maintained. @kcgen and @FeralChild64 can chime in perhaps. I'd say we should remove it altogether if support is sketchy and no one maintains and tests it.

**Comment by kcgen** [comment]
> Just tested with the provided build of dosbox-staging 0.80.1 and am seeing the same issue.
Thanks @AaronBPaden.

Can you try going back more binaries releases - 79, 78, .. , and find when this started?

**Comment by weirddan455** [comment]
> AFAIK, ALSA is heavily deprecated on our side and thus not well maintained. @kcgen and @FeralChild64 can chime in perhaps. I'd say we should remove it altogether if support is sketchy and no one maintains and tests it.
> 
> Then one day I'll switch to RtMidi which supports ALSA, so it might come back for free. https://github.com/thestk/rtmidi

I think @AaronBPaden is talking about MIDI passthrough, which uses ALSA.  If someone has a real MT-32 or something, this is how they would use it with DOSBox.  FluidSynth also has the ability to run as a service and simulate a real MIDI device.  This is how I did it in original DOSBox before discovering Staging.  I don't think this feature should be dropped.

@AaronBPaden Just to clarify, are you saying the FluidSynth service itself starts having problems after running Duke?  As in other programs stop being able to output sound to it?

Also, what distro and FluidSynth version are you running?  I'll try to reproduce later.

**Comment by johnnovak** [comment]
> I think @AaronBPaden is talking about MIDI passthrough, which uses ALSA. If someone has a real MT-32 or something, this is how they would use it with DOSBox. [...] I don't think this feature should be dropped.

Oh yeah, absolutely, MIDI output is a core feature that should never go away. I confused ALSA with OSS, which is indeed deprecated AFAIK... Had another look at RtMidi, and it supports only ALSA and JACK on Linux. So given that JACK is a musician/power-user thing, I assume ALSA is the standard for MIDI output.

So yeah, it definitely stays and should be fixed if it doesn't work. Sorry for the confusion.

**Comment by weirddan455** [comment]
@johnnovak ALSA is the only sound API that the Linux kernel provides.  Everything else (Pulseaudio, Pipewire, JACK) is user-land services that run on top of ALSA.

Pulseaudio and Pipewire are the services you usually talk to for "normal" sound (it's what SDL uses).  They mostly exist to mix sounds from multiple applications, send things to the right sound card, and all that convenience stuff you expect.  I don't think they do MIDI though.  If you're doing MIDI then I think ALSA and JACK are the only options.  As ALSA is built into the kernel, it's available everywhere.

OSS is the old kernel API but it's been deprecated for 20 years or so at this point so don't even worry about that one.

**Comment by johnnovak** [comment]
Thanks for the explanation @weirddan455, today I learned something 😄 I used JACK briefly on macOS for music production, it had some interesting routing capabilities.

**Comment by AaronBPaden** [comment]
After some further investigation, I sort of know why this is happening, though I don't know why dosbox is triggering it and not other software using the alsa sequencer like wine or mido.

Somehow dosbox is causing a page fault kernel side in the alsa sequencer. I guess it can't recover so it just fails to output sound anymore.

@kcgen 

I've tested this with 79 and got the same result... Which makes me think this might be a regression somewhere else in the stack. I'm certain midi was working for me in dosbox-staging in 2022. That said, dosbox is the only software exhibiting this issue, so maybe it's doing something weird.

@weirddan455 

I'm running Arch Linux. Software versions:
Linux 6.5
alsa-lib 1.2.9
pipewire: 0.3.79
wireplumber: 4.14
fluidsynth: 2.3.3

**Comment by johnnovak** [comment]
That's super weird as I don't see how that can happen. I mean, the receiver end, ALSA, should handle all sorts of input, even invalid MIDI sequences, etc. It's basically just a protocol of streaming bytes which constitute the MIDI data, no? So if anything, maybe ALSA is buggy... dunno.

**Comment by weirddan455** [comment]
>  Somehow dosbox is causing a page fault kernel side in the alsa sequencer. I guess it can't recover so it just fails to output sound anymore.

Do you have logs of that?  I'm also on Arch running those same versions so I'll test in a little bit.  I'm on 6.4.12 kernel though.  6.5 is still in the testing repo.  Have you tried with a 6.4.x kernel?

**Comment by AaronBPaden** [comment]
@weirddan455

Here is the log: https://gist.github.com/AaronBPaden/a23a8cd7fba39d351d4837e3737bceb3

Later I will try downgrading the kernel and seeing what happens.

**Comment by weirddan455** [comment]
I tested and everything works fine on my system.  I'd recommend giving this article on the Arch Wiki a read if you haven't already:  https://wiki.archlinux.org/title/FluidSynth

Especially see the note at the bottom.  Would recommend setting `-a pipewire` or `-a pulseaudio` and _not_ `-a alsa` on the FluidSynth daemon.

I will say that log makes me mighty suspicious of that 6.5 kernel though.  Like @johnnovak pointed out, userspace code shouldn't ever be able to cause page faults in kernel space, no matter how buggy it might be.  As 6.5 is less than a week old, you may have found a kernel regression.

**Comment by AaronBPaden** [comment]
I can confirm this seems to be a kernel regression. I get this with 6.5 and 6.5.1 but not 6.4.12. I'm going to do a little more investigation when I get a chance and file a bug at bugzilla.kernel.org.

**Comment by kcgen** [comment]
Great digging, @AaronBPaden. I've renamed and re-labelled the ticket based on latest finding; please feel free to update again as new facts are discovered.

I think we can safely recommend that Linux 6.5 users use `mididevice = fluidsynth`, and ideally our internally compiled, optimized and minimal  fluidsynth library (which has all host APIs disabled).

To get it, use our CI output packages, the release package, or compile with the `--force-fallback-for=fluidsynth` meson setup option. For example:

```shell
meson setup --force-fallback-for=fluidsynth build
```

**Comment by weirddan455** [comment]
6.5.2 just went to the core repo on Arch.  I updated and still cannot reproduce this bug.  @AaronBPaden can you check if it's fixed with the lastest Arch updates?

I don't use the testing repo (Arch is bleeding edge enough without it) so I never tried the 6.5 or 6.5.1 kernels that were reported as problematic.

**Comment by AaronBPaden** [comment]
It's still an issue for me.

Since it's happening kernel side, it would not surprise me to learn that this issue is hardware specific. According to lspci -v I have an `Advanced Micro Devices, Inc. [AMD] Starship/Matisse HD Audio Controller`. I haven't opened a bug at the kernel bugtracker yet. By their rules I need to build Linux from their sources. Makes sense, but I  haven't gotten around to it just yet. And track down whatever is tainting my kernel, probably my mouse drivers...

**Comment by kcgen** [comment]
The Linux 6.5 kernel added MIDI 2.0 to it's ALSA MIDI interface:

https://git.kernel.org/pub/scm/linux/kernel/git/tiwai/sound.git/commit/?h=for-next&id=6b39e30dce18114e3fc27074cee9a2b91a3639d1

Given the kernel devs always avoid breaking userspace, I suspect this is still a regression - perhaps due to the large amount of code changes in the ALSA MIDI module.

**Comment by keenanweaver** [comment]
Just wanted to share my experience if it's of any help. I haven't been able to reproduce the issue. The gist of my setup is:

- NixOS (unstable version), kernel 6.5.2
- Staging in an Arch distrobox (bazzite-arch image, staging version **0.81.0-alpha-1676-gfed6c628d+**, AUR package `dosbox-staging-git`)
- Staging **not** compiled with Fluidsynth disabled
- Fluidsynth systemd user service managed by home-manager
- Real SC-55 hardware
- Also have `Advanced Micro Devices, Inc. [AMD] Starship/Matisse HD Audio Controller`
- Test game was Under a Killing Moon (eXoDOS version)

Setting Dosbox to use alsa & my external Fluidsynth service worked fine, internal Fluidsynth worked fine, SC-55 hardware was fine.

**Comment by kcgen** [comment]
Thanks for the detailed repro tests and feedback @weirddan455 and @keenanweaver.

I will add my results shortly (comparing Kernel 6.3.1 vs. 6.5.3):
```
-rw------- 1 root root  15M Apr 30 17:31 vmlinuz-6.3.1-060301-generic
-rw------- 1 root root  14M Sep 13 01:34 vmlinuz-6.5.3-060503-generic
```

**Comment by AaronBPaden** [comment]
@keenanweaver 

First of all, I'm jealous because you have a real SC-55. :)

Do you have pipewire or pulseaudio installed & can I get your fluidsynth flags/configuration for the user service? I will see if I get the issue with that config.

**Comment by keenanweaver** [comment]
@AaronBPaden I'm jealous of the additional $250 you have that I don't!

I'm using Pipewire. Here's the systemd service as generated by home-manager:

```
[Install]
WantedBy=default.target

[Service]
ExecStart=/nix/store/a5pmfshbvbbvf2c7hm4qqilhk464svzv-fluidsynth-2.3.3/bin/fluidsynth -a pulseaudio -si --gain 0.5 /home/keenan/Music/soundfonts/default.sf2

[Unit]
After=pipewire-pulse.service
BindsTo=pipewire-pulse.service
Description=FluidSynth Daemon
Documentation=man:fluidsynth(1)
```

**Comment by kcgen** [comment]
Can't repro on Ubuntu 24.04 running Linux server 6.5.3-060503-generic #202309130834 SMP PREEMPT_DYNAMIC

First launch `qsynth` connected via pipewire's pulse-audio interface.

Then launched with `--noprimaryconf` plus local dosbox.conf:

```ini
[midi]
mididevice = alsa
```

Left qsynth running and didn't touch its GUI. Then did multiple restarts of DOSBox Staging. The MIDI audio from qsynth always started up properly and reported activity. 

![2023-09-15_08-50](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/d65a1d3d-750b-47fa-945f-2b57d4576b5e)

Given three separate users using 3 separate Linux OSes have not been able to reproduce it all with Linux 6.5.x, I'm going to close this as "can't reproduce". 

Unfortunately the data is pointing to a one-off system-specific issue on your side, @AaronBPaden.  Maybe it's not (could be a kernel but that only manifests on your model of hardware or configuration?); but unfortunately I think we're at the end of the line on this ticket.

That said - if there's any news for future readers, please keep updating it!

**Comment by AaronBPaden** [comment]
It seems this was fixed as of at least kernel 6.5.4
